File: src/ts/ssh-tcp/services/channelForwarder.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as net from 'net';
import {
	SshChannel,
	SshChannelClosedEventArgs,
	Trace,
	TraceLevel,
	SshTraceEventIds,
} from '@microsoft/dev-tunnels-ssh';
import { CancellationTokenSource, Disposable } from 'vscode-jsonrpc';
import { PortForwardingService } from './portForwardingService';

export class ChannelForwarder implements Disposable {
	private readonly disposeCancellationSource = new CancellationTokenSource();

	/* @internal */
	public constructor(
		private readonly pfs: PortForwardingService,
		public readonly channel: SshChannel,
		public readonly socket: net.Socket,
	) {
		socket.on('data', this.onSocketDataReceived.bind(this));
		socket.on('error', this.onSocketError.bind(this));
		socket.on('close', this.onSocketClosed.bind(this));
		channel.onDataReceived(this.onChannelDataReceived.bind(this));
		channel.onClosed(this.onChannelClosed.bind(this));
	}

	private get trace(): Trace {
		return this.channel.session.trace;
	}

	private onSocketDataReceived(data: Buffer): void {
		this.socket.pause(); // Block further data events while sending.
		this.channel
			.send(data, this.disposeCancellationSource.token)
			.then(this.socket.resume.bind(this.socket), (e: Error) => {
				this.trace(
					TraceLevel.Warning,
					SshTraceEventIds.portForwardConnectionFailed,
					`Forwarder error sending channel data: ${e.message}`,
					e,
				);
			});
	}

	private onSocketClosed(hadError: boolean): void {
		const closePromise = hadError
			? this.channel.close('SIGABRT', 'Socket closed with error.')
			: this.channel.close();
		closePromise.catch((e: Error) => {
			this.trace(
				TraceLevel.Error,
				SshTraceEventIds.portForwardConnectionFailed,
				`Forwarder channel close failed with error: ${e.message}`,
				e,
			);
		});
	}

	private onSocketError(error: Error): void {
		this.trace(
			TraceLevel.Info,
			SshTraceEventIds.portForwardChannelClosed,
			`Forwarder socket closed with error: ${error.message}`,
		);
		this.channel.close('SIGABRT', error.message).catch((e: Error) => {
			this.trace(
				TraceLevel.Error,
				SshTraceEventIds.portForwardConnectionFailed,
				`Forwarder channel close failed with error: ${e.message}`,
				e,
			);
		});
	}

	private onChannelDataReceived(data: Buffer): void {
		this.socket.write(data, (error?: Error) => {
			if (error) {
				this.trace(
					TraceLevel.Error,
					SshTraceEventIds.portForwardConnectionFailed,
					`Forwarder failed to write to socket: ${error.message}`,
					error,
				);
			} else {
				this.channel.adjustWindow(data.length);
			}
		});
	}

	private onChannelClosed(e: SshChannelClosedEventArgs): void {
		if (!e.errorMessage) {
			this.trace(
				TraceLevel.Info,
				SshTraceEventIds.portForwardConnectionClosed,
				`Forwarder channel ${this.channel.channelId} closed.`,
			);
			this.socket.end();
		} else {
			this.trace(
				TraceLevel.Info,
				SshTraceEventIds.portForwardConnectionClosed,
				`Forwarder channel ${this.channel.channelId} closed with error: ${e.errorMessage}`,
			);
			// TODO: Destroy the socket in a way that causes a connection reset:
			// https://github.com/nodejs/node/issues/27428
			this.socket.destroy();
		}

		const index = this.pfs.channelForwarders.indexOf(this);
		if (index >= 0) {
			this.pfs.channelForwarders.splice(index, 1);
		}

		this.dispose();
	}

	public dispose(): void {
		this.disposeCancellationSource.dispose();
		this.socket.destroy();
	}
}

File: src/cs/Ssh.Tcp/Services/ChannelForwarder.cs
﻿// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Events;

namespace Microsoft.DevTunnels.Ssh.Tcp;

internal class ChannelForwarder : IDisposable
{
	private readonly PortForwardingService pfs;
	private bool disposed;
	private readonly SemaphoreSlim receiveSemaphore;
	private readonly ConcurrentQueue<Buffer> receiveQueue;
	private SshChannelClosedEventArgs? channelClosedEvent;
	private readonly NetworkStream stream;
	private readonly TraceSource trace;
	private readonly CancellationTokenSource disposeCancellationSource;

	public ChannelForwarder(PortForwardingService pfs, SshChannel channel, TcpClient client)
	{
		this.pfs = pfs ?? throw new ArgumentNullException(nameof(pfs));
		Channel = channel ?? throw new ArgumentNullException(nameof(channel));
		Client = client ?? throw new ArgumentNullException(nameof(client));
		this.stream = client.GetStream() ?? throw new ArgumentNullException(nameof(client));

		this.receiveSemaphore = new SemaphoreSlim(0);
		this.receiveQueue = new ConcurrentQueue<Buffer>();
		this.trace = channel.Session.Trace;
		this.disposeCancellationSource = new CancellationTokenSource();

		Channel.DataReceived += OnChannelDataReceived;
		Channel.Closed += OnChannelClosed;

		ForwardFromStreamToChannel(new Buffer(4096));
		ForwardFromChannelToStream();
	}

	public SshChannel Channel { get; }

	public TcpClient Client { get; set; }

	private void OnChannelDataReceived(object? sender, Buffer data)
	{
		if (this.disposed)
		{
			return;
		}

		// Enqueue a copy of the buffer because the current one may be re-used by the caller.
		var copy = new Buffer(data.Count);
		data.CopyTo(copy);

		this.receiveQueue.Enqueue(copy);

		try
		{
			this.receiveSemaphore.Release();
		}
		catch (ObjectDisposedException)
		{
			// The semaphore was disposed.
		}

		Channel.AdjustWindow((uint)data.Count);
	}

	private void OnChannelClosed(object? sender, SshChannelClosedEventArgs e)
	{
		if (this.disposed)
		{
			return;
		}

		this.channelClosedEvent = e;

		try
		{
			this.receiveSemaphore.Release();
		}
		catch (ObjectDisposedException)
		{
			// The semaphore was disposed.
		}
	}

	private async void ForwardFromChannelToStream()
	{
		try
		{
			bool forwarding;
			do
			{
				forwarding = await ForwardFromChannelToStreamAsync().ConfigureAwait(false);
			}
			while (forwarding);

			this.pfs.RemoveChannelForwarder(this);
			this.Dispose();
		}
		catch (Exception ex)
		{
			// Catch all exceptions in this async void method.
			this.trace.TraceEvent(
				TraceEventType.Error,
				SshTraceEventIds.UnknownError,
				$"{nameof(PortForwardingService)} unexpected error reading channel stream: {ex}");
		}
	}

	private async Task<bool> ForwardFromChannelToStreamAsync()
	{
		try
		{
			await this.receiveSemaphore.WaitAsync(this.disposeCancellationSource.Token)
				.ConfigureAwait(false);
		}
		catch (ObjectDisposedException)
		{
			// The semaphore was disposed.
			CloseStream(abort: true);
			return false;
		}
		catch (OperationCanceledException)
		when (this.disposeCancellationSource.IsCancellationRequested)
		{
			// The channel-forwarder was disposed.
			CloseStream(abort: true);
			return false;
		}

		if (this.receiveQueue.TryDequeue(out Buffer data))
		{
			try
			{
#if SSH_ENABLE_SPAN
				await this.stream.WriteAsync(
					data.Memory, this.disposeCancellationSource.Token).ConfigureAwait(false);
#else
				await this.stream.WriteAsync(
					data.Array,
					data.Offset,
					data.Count,
					this.disposeCancellationSource.Token).ConfigureAwait(false);
#endif
				return true;
			}
			catch (OperationCanceledException)
			when (this.disposeCancellationSource.IsCancellationRequested)
			{
				// The channel-forwarder was disposed.
				CloseStream(abort: true);
				return false;
			}
		}
		else
		{
			if (this.channelClosedEvent != null)
			{
				var errorMessage = this.channelClosedEvent.ErrorMessage
					?? this.channelClosedEvent.Exception?.Message;
				var message = string.IsNullOrEmpty(errorMessage) ?
					$"Forwarder channel {Channel.ChannelId} closed." :
					$"Forwarder channel {Channel.ChannelId} closed with error: " + errorMessage;
				this.trace.TraceEvent(
					TraceEventType.Information,
					SshTraceEventIds.PortForwardChannelClosed,
					message);

				CloseStream(abort: (errorMessage != null));
			}

			// Reached end of stream.
			return false;
		}
	}

	private void CloseStream(bool abort)
	{
		try
		{
			if (abort)
			{
				var socket = Client.Client;
				socket.Abort();
			}
			else
			{
				this.stream.Close();
			}
		}
		catch (Exception ex)
		{
			this.trace.TraceEvent(
				TraceEventType.Warning,
				SshTraceEventIds.UnknownError,
				$"{nameof(PortForwardingService)} unexpected error clising connection: {ex}");
			return;
		}

		this.trace.TraceEvent(
			TraceEventType.Information,
			SshTraceEventIds.PortForwardChannelClosed,
			$"Channel forwarder {(abort ? "aborted" : "closed")} connection.");
	}

	private async void ForwardFromStreamToChannel(Buffer buffer)
	{
		try
		{
			bool forwarding;
			do
			{
				forwarding = await ForwardFromStreamToChannelAsync(
					buffer, this.disposeCancellationSource.Token).ConfigureAwait(false);
			}
			while (forwarding);

			this.pfs.RemoveChannelForwarder(this);
			this.Dispose();
		}
		catch (Exception ex)
		{
			// Catch all exceptions in this async void method.
			this.trace.TraceEvent(
				TraceEventType.Error,
				SshTraceEventIds.UnknownError,
				$"{nameof(PortForwardingService)} unexpected error reading channel stream: {ex}");
		}
	}

	private async Task<bool> ForwardFromStreamToChannelAsync(
		Buffer buffer, CancellationToken cancellation)
	{
		int count;
		Exception? ex = null;
		try
		{
#if NETSTANDARD2_0 || NET4
			count = await this.stream.ReadAsync(
				buffer.Array, buffer.Offset, buffer.Count, cancellation).ConfigureAwait(false);
#else
			count = await this.stream.ReadAsync(buffer.Memory, cancellation).ConfigureAwait(false);
#endif
		}
		catch (IOException ioex)
		{
			ex = ioex;
			count = 0;
		}
		catch (SocketException sockex)
		{
			ex = sockex;
			count = 0;
		}
		catch (OperationCanceledException) when (cancellation.IsCancellationRequested)
		{
			return false;
		}

		// Do not use the (dispose) cancellation token when writing to the channel, because
		// an interrupted write can cause the whole SSH session to disconnect.
		if (count > 0)
		{
			await Channel.SendAsync(buffer.Slice(0, count), CancellationToken.None)
				.ConfigureAwait(false);
			return true;
		}
		else if (ex == null)
		{
			string message = "Channel forwarder reached end of stream.";
			this.trace.TraceEvent(TraceEventType.Verbose, SshTraceEventIds.ChannelClosed, message);
			try
			{
				await Channel.SendAsync(Buffer.Empty, CancellationToken.None).ConfigureAwait(false);
				await Channel.CloseAsync(CancellationToken.None).ConfigureAwait(false);
			}
			catch (OperationCanceledException)
			{
			}
		}
		else
		{
			string message = $"Channel forwarder stream read error: {ex.Message}";
			this.trace.TraceEvent(TraceEventType.Verbose, SshTraceEventIds.ChannelClosed, message);
			try
			{
				await Channel.CloseAsync("SIGABRT", ex.Message, CancellationToken.None)
					.ConfigureAwait(false);
			}
			catch (OperationCanceledException)
			{
			}
		}

		return false;
	}

	/// <summary>
	/// Diposes the service; called when the session is disposing.
	/// </summary>
	public void Dispose()
	{
		Dispose(true);
		GC.SuppressFinalize(this);
	}

	/// <summary>
	/// Subclasses may override this method to dispose any resources.
	/// </summary>
	/// <param name="disposing">True if managed objects are disposed.</param>
	protected virtual void Dispose(bool disposing)
	{
		if (disposing && !this.disposed)
		{
			this.disposed = true;
			try
			{
				this.disposeCancellationSource.Cancel();
			}
			catch (ObjectDisposedException) { }
			this.disposeCancellationSource.Dispose();

			this.receiveSemaphore.Dispose();
		}
	}
}
